caffe2/contrib/tensorrt/tensorrt_tranformer.cc

```diff
@@ -0,0 +1,423 @@
+  for (const auto& op: init_net.op()) {
+    CAFFE_ENFORCE(op.output_size() == 1);
+    auto it = initialization_list.find(op.output(0));
+    if (it != initialization_list.end()) {
+      auto* init_tensor = model->mutable_graph()->add_initializer();
+      exporter.InitOpToTensorProto(op, init_tensor);
+      initialization_list.erase(it);
+    }
+  }
+  CAFFE_ENFORCE(initialization_list.empty(), "Unfulfilled initialization list");
+  for (const auto& t: model->graph().initializer()) {
+    LOG(INFO) << "Initializer: " << t.name();
+  }
+
+  // Onnx model is ready. Call onnx-trt to convert to one trt c2 op
+  std::string model_str;
+  model->SerializeToString(&model_str);
+  new_ops->emplace_back(BuildTrtOp(model_str, output_shape_hints));
+
+  model->mutable_graph()->clear_node();
+}
+
+// Cutting off the runnable part and replace with tensor ops. Asssume the nets
+// were topologically sorted
+void TensorRTTransformer::Transform(
+    NetDef* init_net,
+    NetDef* pred_net,
+    const std::unordered_map<std::string, TensorShape>& input_shape_hints) {
+  const auto shape_hints = InferShapes(init_net, pred_net, input_shape_hints);
+
+  std::unordered_set<std::string> weights;
+  if (init_net) {
+    for (const auto& op : init_net->op()) {
+      CAFFE_ENFORCE(op.type().find("GivenTensor") == 0);
+      CAFFE_ENFORCE(op.type().rfind("Fill") == op.type().size() - 4);
+      CAFFE_ENFORCE(op.output_size() == 1);
+      for (const auto& op_output : op.output()) {
+        weights.emplace(op_output);
+      }
+    }
+  }
+
+  CAFFE_ENFORCE(pred_net, "pred_net cannot be nullptr");
+  ::ONNX_NAMESPACE::ModelProto onnx_model;
+  FillModelInfo(&onnx_model);
+
+  std::vector<OperatorDef> new_ops;
+  bool trt_group = false;
+  auto importer = InferObject(onnx2trt::createImporter(nullptr));
+  int op_idx = 0;
+  int start = 0;
+  int end = 0;
+  for (const OperatorDef& op : pred_net->op()) {
+    bool support_trt = true;
+    const OpSchema* schema = OpSchemaRegistry::Schema(op.type());
+    caffe2::onnx::ConvertedResult results;
+    if (!schema or schema->onnx_schema().empty()) {
+      LOG(INFO) << "Cannot export c2 op " << op.type() << " to onnx";
+      support_trt = false;
+    } else {
+      // One c2 op can be converted into multiple onnx nodes. For simplicity, we
+      // enforce all or nothing here
+      results = onnx::OnnxExporter().Caffe2OpToOnnxNodes(op, shape_hints);
+      for (const auto& n : results.first) {
+        if (!importer->supports(n)) {
+          LOG(INFO) << "TRT does not support ONNX node " << n.op_type();
+          support_trt = false;
+          break;
+        }
+      }
+    }
+
+    if (support_trt) {
+      const auto& node_protos = results.first;
+      if (!trt_group) {
+        trt_group = true;
+        start = op_idx;
+      }
+      for (const auto& n : node_protos) {
+        auto* node = onnx_model.mutable_graph()->add_node();
+        node->CopyFrom(n);
+      }
+    } else {
```

Comment:
I don't understand the logic here. Why convert cluster to TRT can be successful if an op is not supported?

Reply:
If we cannot convert some ops to trt op, due to 1. We cannot convert c2 op to onnx op
2. Trt backend doesn't support the onnx op
, It is still ok, we just leave that c2 op as is because we can still run it in c2.